src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -386,34 +410,61 @@ public override int GetHashCode()
-            TypeWithAnnotations oldElementType = ElementTypeWithAnnotations;
-            TypeWithAnnotations newElementType;
-
-            if (!oldElementType.ApplyNullableTransforms(defaultTransformFlag, transforms, ref position, out newElementType))
+            var elementType = ElementTypeWithAnnotations;
+            while (elementType.Type is ArrayTypeSymbol arrayType)
             {
-                result = this;
-                return false;
+                elementType.AddNullableTransformShallow(transforms);
+                elementType = arrayType.ElementTypeWithAnnotations;
             }
 
-            result = WithElementType(newElementType);
-            return true;
+            elementType.AddNullableTransforms(transforms);
         }
 
-        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
-        {
-            return WithElementType(transform(ElementTypeWithAnnotations));
-        }
+        internal override TypeSymbol ApplyNullableTransforms(NullableTransformStream stream) =>
+            TransformNullable(ApplyTransformNullableTransform.Instance, stream);
+
+        internal override TypeSymbol SetObliviousNullabilityForReferenceTypes() =>
+            TransformNullable(SetObliviousNullableTransform.Instance, null);
 
         internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind variance)
         {
             Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
-            TypeWithAnnotations elementType = ElementTypeWithAnnotations.MergeNullability(((ArrayTypeSymbol)other).ElementTypeWithAnnotations, variance);
-            return WithElementType(elementType);
+            return TransformNullable(MergeNullableTransform.Instance, ((ArrayTypeSymbol)other, variance));
+        }
+
+        private ArrayTypeSymbol TransformNullable<TState, TInitialState>(NullableTransform<TState, TInitialState> transform, TInitialState initialState)
+        {
+            // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+            // recursive solution here. Need to take an iterative approach to building up the annotations here.
+            var builder = ArrayBuilder<TState>.GetInstance();
+
+            // Dig through the element types to get the set of ArrayTypeSymbol instances that need to be marked
+            // as oblivious
+            var array = this;
+            var state = transform.GetInitialState(initialState, this);
+            builder.Push(state);
+            do
+            {
+                array = array.ElementType as ArrayTypeSymbol;
+                if (array is object)
+                {
+                    state = transform.GetState(state, array);
+                    builder.Push(state);
+                }
+            }
+            while (array is object);
```

Comment:
>array is object [](start = 19, length = 15)

We already checked this condition above. Consider using ```while (true)``` and breaking from the ```else``` for the ```if``` above.